Userland/Libraries/LibVirtGPU/CommandBufferBuilder.cpp

```diff
@@ -0,0 +1,310 @@
+    builder.appendu32(height);           // height
+    builder.appendu32(depth);            // depth
+    builder.appendu32(0);                // data_offset
+    builder.appendu32(direction);        // direction
+}
+
+void CommandBufferBuilder::append_end_transfers_3d()
+{
+    CommandBuilder builder(m_buffer, Protocol::VirGLCommand::END_TRANSFERS, 0);
+}
+
+void CommandBufferBuilder::append_draw_vbo(Protocol::PipePrimitiveTypes primitive_type, u32 count)
+{
+    CommandBuilder builder(m_buffer, Protocol::VirGLCommand::DRAW_VBO, 0);
+    builder.appendu32(0);                             // start
+    builder.appendu32(count);                         // count
+    builder.appendu32(to_underlying(primitive_type)); // mode
+    builder.appendu32(0);                             // indexed
+    builder.appendu32(1);                             // instance_count
+    builder.appendu32(0);                             // index_bias
+    builder.appendu32(0);                             // start_instance
+    builder.appendu32(0);                             // primitive_restart
+    builder.appendu32(0);                             // restart_index
+    builder.appendu32(0);                             // min_index
+    builder.appendu32(0xffffffff);                    // max_index
+    builder.appendu32(0);                             // cso
+}
+
+void CommandBufferBuilder::append_gl_clear(float r, float g, float b, float a)
+{
+    CommandBuilder builder(m_buffer, Protocol::VirGLCommand::CLEAR, 0);
+    Protocol::ClearType clear_flags {};
+    clear_flags.flags.color0 = 1;
+    builder.appendu32(clear_flags.value);
+    builder.appendf32(r);
+    builder.appendf32(g);
+    builder.appendf32(b);
+    builder.appendf32(a);
+    builder.appendf64(1.0);
+    builder.appendu32(0);
+}
+
+void CommandBufferBuilder::append_gl_clear(double depth)
+{
+    CommandBuilder builder(m_buffer, Protocol::VirGLCommand::CLEAR, 0);
+    Protocol::ClearType clear_flags {};
+    clear_flags.flags.depth = 1;
+    builder.appendu32(clear_flags.value);
+    builder.appendf32(0.f);
+    builder.appendf32(0.f);
+    builder.appendf32(0.f);
+    builder.appendf32(0.f);
+    builder.appendf64(depth);
+    builder.appendu32(0);
+}
+
+void CommandBufferBuilder::append_set_vertex_buffers(u32 stride, u32 offset, Protocol::ResourceID resource)
+{
+    CommandBuilder builder(m_buffer, Protocol::VirGLCommand::SET_VERTEX_BUFFERS, 0);
+    builder.appendu32(stride);
+    builder.appendu32(offset);
+    builder.appendu32(resource.value());
+}
+
+void CommandBufferBuilder::append_create_blend(Protocol::ObjectHandle handle)
+{
+    CommandBuilder builder(m_buffer, Protocol::VirGLCommand::CREATE_OBJECT, to_underlying(Protocol::ObjectType::BLEND));
+    builder.appendu32(handle.value());
+    builder.appendu32(4); // Enable dither flag, and nothing else
+    builder.appendu32(0);
+    builder.appendu32(0x78000000); // Enable all bits of color mask for color buffer 0, and nothing else
```

Comment:
Can we rewrite these two magic values `4` and `0x78000000` so it becomes clear what they're setting? (So they don't need a comment)